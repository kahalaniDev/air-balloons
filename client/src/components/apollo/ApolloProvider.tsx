import React from "react";
import { MockedProvider } from "@apollo/client/testing";
import { ApolloProvider as ApolloNativeProvider } from "@apollo/client";
import { apolloClient } from "../../infrastructure/apollo/client";
import { loginGraphqlMock } from "../../features/account/api/login/mock/loginGraphqlMock";
import { getBalloonsGraphqlMock } from "../../features/balloons/api/getBalloons/mock/getBalloonsGraphqlMock";
import { getBalloonGraphqlMock } from "../../features/balloons/api/getBalloon/mock/getBalloonGraphqlMock";
import { addBalloonGraphqlMock } from "../../features/balloons/api/addBalloon/mock/addBalloonGraphqlMock";

interface Props {
  useMocks?: boolean;
}

const ApolloProvider: React.FC<Props> = ({ useMocks, children }) => {
  if (useMocks)
    return (
      <MockedProvider
        mocks={[
          ...loginGraphqlMock,
          ...getBalloonsGraphqlMock,
          ...getBalloonGraphqlMock,
          ...addBalloonGraphqlMock,
        ]}
      >
        <>{children}</>
      </MockedProvider>
    );
  return (
    <ApolloNativeProvider client={apolloClient}>
      <>{children}</>
    </ApolloNativeProvider>
  );
};

export default ApolloProvider;
