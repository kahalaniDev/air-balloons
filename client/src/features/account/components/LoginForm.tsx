import React, { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { useApolloClient } from "@apollo/client";
import { useFormik } from "formik";
import useAppDispatch from "../../../hooks/useAppDispatch";
import useAppSelector from "../../../hooks/useAppSelector";
import { login, resetError } from "../slices/accountSlice";
import { loginGraphql } from "../api/login/loginGraphql";
import { loginRest } from "../api/login/loginRest";
import { validationLoginSchema } from "../utils/validators";
import { SERVER_TYPE, ACTIVE_SERVER } from "../../../infrastructure/config";
import { IUserCredentials } from "../models/interfaces";
import FormError from "../../../components/form/FormError";
import { Box, TextField } from "@mui/material";
import FormButton from "../../../components/form/FormButton";

type Props = {};

const formErrorCodes = [401];

const LoginForm: React.FC<Props> = () => {
  const client = useApolloClient();
  const [popupOpen, setPopupOpen] = useState(false);

  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { loading, error } = useAppSelector((state) => ({
    error: state.account.error,
    loading: state.account.loading,
  }));

  const formik = useFormik({
    initialValues: {
      username: "",
      password: "",
    },
    validationSchema: validationLoginSchema,
    onSubmit: async (userCred) => {
      const resultAction = await dispatch(
        login({
          userCred,
          loginRequest:
            ACTIVE_SERVER === SERVER_TYPE.GRAPHQL
              ? (userCredentials: IUserCredentials) =>
                  loginGraphql(client, userCredentials)
              : loginRest,
        })
      );

      if (login.fulfilled.match(resultAction)) navigate("/");
      else if (!formErrorCodes.includes(resultAction.payload!.statusCode))
        setPopupOpen(true);
    },
  });

  const isFormValuesChanged = () => {
    return (
      formik.initialValues.username !== formik.values.username ||
      formik.initialValues.password !== formik.values.password
    );
  };

  useEffect(() => {
    return () => {
      dispatch(resetError());
    };
  }, [dispatch]);

  return (
    <Box
      component="form"
      noValidate
      sx={{ mt: 1 }}
      onSubmit={formik.handleSubmit}
    >
      <TextField
        margin="normal"
        disabled={loading}
        {...formik.getFieldProps("username")}
        error={formik.touched.username && Boolean(formik.errors.username)}
        helperText={formik.touched.username && formik.errors.username}
        fullWidth
        id="username"
        label="Username"
        autoComplete="username"
        autoFocus
      />
      <TextField
        margin="normal"
        disabled={loading}
        {...formik.getFieldProps("password")}
        error={formik.touched.password && Boolean(formik.errors.password)}
        helperText={formik.touched.password && formik.errors.password}
        fullWidth
        label="Password"
        type="password"
        id="password"
        autoComplete="current-password"
      />
      <FormButton
        title="Sign In"
        loading={loading}
        disabled={!isFormValuesChanged() || !formik.isValid}
      />
      <FormError
        popupOpen={popupOpen}
        closePopup={() => setPopupOpen(false)}
        error={error}
        formErrorCodes={formErrorCodes}
      />
    </Box>
  );
};

export default LoginForm;
