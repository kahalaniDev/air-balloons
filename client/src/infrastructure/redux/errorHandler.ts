import { ApolloError, isApolloError } from "@apollo/client";
import axios, { AxiosError } from "axios";
import { GraphQLError } from "graphql";
import { IAppError } from "../../models/interfaces";

const GLOBAL_STATUS_CODE = 500;
const INTERNAL_ERR_MSG = "Internal server problem, please try again";
const NEWTWORK_ERR_MSG = "Unable to access the server, please try again";

export const errorHandler = (err: unknown) => {
  if (axios.isAxiosError(err)) return axiosErrorBuilder(err);
  if (err instanceof Error && isApolloError(err))
    return apolloErrorBuilder(err);
  return {
    statusCode: 500,
    message: INTERNAL_ERR_MSG,
  };
};

const axiosErrorBuilder = (axiosError: AxiosError) => {
  console.log(axiosError);
  const statusText = axiosError.response?.statusText
    ? axiosError.response!.statusText
    : "";
  const statusCode = axiosError.response!.status;
  switch (statusText) {
    case "Not Found":
      return { statusCode, message: axiosError.response!.data };
    case "Bad Request":
      return { statusCode, message: axiosError.response!.data };
    case "Conflict":
      return { statusCode, message: axiosError.response!.data };
    case "Unauthorized":
      return { statusCode, message: axiosError.response!.data };
    default:
      return { statusCode, message: INTERNAL_ERR_MSG };
  }
  // return {
  //   statusCode: axiosError.response?.status
  //     ? axiosError.response?.data?.statusCode
  //     : axiosError.code,
  //   message: axiosError.response?.data?.message
  //     ? axiosError.response?.data?.message
  //     : INTERNAL_ERR_MSG,
  // };
};

const apolloErrorBuilder = (apolloError: ApolloError): IAppError => {
  if (apolloError.networkError)
    return { statusCode: GLOBAL_STATUS_CODE, message: NEWTWORK_ERR_MSG };
  if (apolloError.graphQLErrors)
    return graphqlErrorBuilder(apolloError.graphQLErrors[0]);
  return { statusCode: GLOBAL_STATUS_CODE, message: INTERNAL_ERR_MSG };
};

const graphqlErrorBuilder = (graphQLError: GraphQLError): IAppError => {
  const statusCode = graphQLError.extensions.codeNum as number;
  switch (graphQLError.extensions.code) {
    case "UNAUTHENTICATED":
    case "BAD_USER_INPUT":
      return { statusCode, message: graphQLError.message };
    default:
      return { statusCode, message: INTERNAL_ERR_MSG };
  }
};
