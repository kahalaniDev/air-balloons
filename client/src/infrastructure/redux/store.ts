import { configureStore, PreloadedState } from "@reduxjs/toolkit";
import accountReducer, {
  initialState as accountInitialState,
} from "../../features/account/slices/accountSlice";
import balloonsReducer, {
  initialState as balloonInitialState,
} from "../../features/balloons/slices/balloonsSlice";
import { isLoggedIn } from "../../features/account/slices/accountSlice";

type InitialState = {
  account?: typeof accountInitialState;
  balloon?: typeof balloonInitialState;
};

export const createStore = (preloadedState?: PreloadedState<InitialState>) =>
  configureStore({
    preloadedState,
    reducer: { account: accountReducer, balloons: balloonsReducer },
  });

const store = createStore();
store.dispatch(isLoggedIn());

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
