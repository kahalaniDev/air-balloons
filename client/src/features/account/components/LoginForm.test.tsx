import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router";
import ApolloProvider from "../../../components/apollo/ApolloProvider";
import LoginForm from "./LoginForm";
import { Provider as ReduxProvider } from "react-redux";
import { createStore } from "../../../infrastructure/redux/store";

jest.mock("../../../infrastructure/config.ts", () => ({
  __esModule: true,
  USE_MOCK_SERVER: true,
  SERVER_TYPE: {
    GRAPHQL: 0,
    REST: 1,
  },
  ACTIVE_SERVER: 0,
}));

describe("LoginForm", () => {
  test("check if LoginForm match to snapshot", () => {
    expect(
      render(
        <MemoryRouter>
          <ApolloProvider useMocks>
            <ReduxProvider store={createStore()}>
              <LoginForm />
            </ReduxProvider>
          </ApolloProvider>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });
  test("validate empty inputs and showing error", async () => {
    render(
      <MemoryRouter>
        <ApolloProvider useMocks>
          <ReduxProvider store={createStore()}>
            <LoginForm />
          </ReduxProvider>
        </ApolloProvider>
      </MemoryRouter>
    );
    userEvent.click(screen.getByRole("button", { name: /SIGN IN/i }));
    expect(await screen.findByText("Username is required")).toBeInTheDocument();
    expect(await screen.findByText("Password is required")).toBeInTheDocument();
  });

  test("check login button click with wrong credentials", async () => {
    render(
      <MemoryRouter>
        <ApolloProvider useMocks>
          <ReduxProvider store={createStore()}>
            <LoginForm />
          </ReduxProvider>
        </ApolloProvider>
      </MemoryRouter>
    );
    userEvent.type(screen.getByLabelText(/Username/i), "beni");
    userEvent.type(screen.getByLabelText(/Password/i), "!1234567");
    expect(screen.getByRole("button", { name: /SIGN IN/i })).toBeEnabled();
    userEvent.click(screen.getByRole("button", { name: /SIGN IN/i }));
    expect(
      await screen.findByText(/incorrect username or password/i)
    ).toBeInTheDocument();
  });

  test("check login button click with right credentials", async () => {
    render(
      <MemoryRouter>
        <ApolloProvider useMocks>
          <ReduxProvider store={createStore()}>
            <LoginForm />
          </ReduxProvider>
        </ApolloProvider>
      </MemoryRouter>
    );
    userEvent.type(screen.getByLabelText(/Username/i), "daniel");
    userEvent.type(screen.getByLabelText(/Password/i), "!1234567");
    expect(screen.getByRole("button", { name: /SIGN IN/i })).toBeEnabled();
    userEvent.click(screen.getByRole("button", { name: /SIGN IN/i }));
    await waitFor(() =>
      expect(
        screen.queryByText(/incorrect username or password/i)
      ).not.toBeInTheDocument()
    );
  });
});
