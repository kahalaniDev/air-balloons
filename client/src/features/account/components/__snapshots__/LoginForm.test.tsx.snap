// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm check if LoginForm match to snapshot 1`] = `
<DocumentFragment>
  <form
    class="MuiBox-root css-164r41r"
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="username"
        id="username-label"
      >
        Username
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="username"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="username"
          name="username"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1z7n62"
          >
            <span>
              Username
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="password"
        id="password-label"
      >
        Password
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="current-password"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="password"
          name="password"
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-173wfxe"
          >
            <span>
              Password
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiBox-root css-1koepnq"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButtonBase-root  css-17k1s6h-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Sign In
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
